### DISCUSSION TITLE
Ping cause packets to arrive at very slow pace

### DISCUSSION THREAD

**Discussion by Ahmed310** [jchristn/WatsonTcp#173] [Q&A]
Not related to WatsopTcp specifically, It is happening to other websockets library as well. 

While running app on real server or when I emulate ping on local machine message arrives at very slow paces even at 100ms ping. My question is am is missing something or do I need to disable Nagle or something like that.

In case of UDP ping caused packets to arrived at same speed but overall packets are delayed by ping Whereas in TCP each message is arriving at slow pace as seen in video below

![ping_delay](https://user-images.githubusercontent.com/8081578/145575773-5101816e-4253-440b-8f36-c96045e2f451.gif)

**Comment by jchristn** [comment]
Hi @Ahmed310 thanks for your post.  Is clumsy actually doing a ping or introducing delay?  It appears to be the latter.  If so, I'd want to better understand how the apps sending and receiving messages are built.  Are the send operations happening in parallel?  Is there some kind of back-and-forth, i.e. a send-and-wait behavior?  From the image it's clearly problematic that counts are dropping so dramatically.  I imagine "count" is number of messages received/handled?  Looks like it drops to 1-2 per second when 50ms of lag is introduced (in your case, 100ms total, 50ms in, 50ms out).

**Comment by Ahmed310** [comment]
not sure how clumsy works but here is the link [ http://jagt.github.io/clumsy/index.html ] 

- there is one dedicated thread for networking on network msg received it enqueued into  receive msgs channel (https://www.nuget.org/packages/System.Threading.Channels/ ) and a process function is dequeuing them from main thread and sending it to app, similarly there is  one channel for sent msgs main thread puts data in it and networking thread is call send function after reading data from sent channel
- count is the number of messages a server received from clients, whenever client ping's server the count gets incremented and sent to client


Question: Is Nagle algorithm is disabled in WatsonTcp ? if not is there a way to disable it    ( https://docs.microsoft.com/en-us/dotnet/api/system.net.sockets.socket.nodelay?view=net-6.0 )

**Reply by jchristn** [reply]
It is not modified, and the default is ```false```.  I can modify the settings class to allow this to be set prior to connect.  Give me a moment.

**Reply by jchristn** [reply]
Try v4.8.14.5, the ```NoDelay``` property is hanging off of the ```Settings``` object for both ```WatsonTcpClient``` and ```WatsonTcpServer```.  The default is ```false```.
NuGet: https://www.nuget.org/packages/WatsonTcp/4.8.14.5
Commit: https://github.com/jchristn/WatsonTcp/commit/2a99eaa096f3f1604ea134a00880e28e08dce560

Also I added you to the README as someone who made the library better.  Thanks for your help @Ahmed310 !!

**Reply by Ahmed310** [reply]
that's awesome and you are quick to response really appreciated it, keep up the good work
